From 703ee352aad336bfabc8bf05bc9248a346a0cf26 Mon Sep 17 00:00:00 2001
From: Subodh Iyengar <subodh@fb.com>
Date: Tue, 27 Jan 2015 09:43:37 -0800
Subject: [PATCH] modifications to openssl 1.0.2

Summary:
modifications to openssl 1.0.2's build system
so that it will compile with reduced options.
---
 Makefile.org        |  4 ++--
 Makefile.shared     |  4 ++--
 conf                | 59 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 crypto/evp/digest.c |  6 ++++++
 4 files changed, 69 insertions(+), 4 deletions(-)
 create mode 100755 conf

diff --git a/Makefile.org b/Makefile.org
index b7a3f96..06fb325 100644
--- a/Makefile.org
+++ b/Makefile.org
@@ -136,9 +136,9 @@ FIPSCANLIB=
 
 BASEADDR=
 
-DIRS=   crypto ssl engines apps test tools
+DIRS=   crypto
 ENGDIRS= ccgost
-SHLIBDIRS= crypto ssl
+SHLIBDIRS= crypto
 
 # dirs in crypto to build
 SDIRS=  \
diff --git a/Makefile.shared b/Makefile.shared
index e753f44..cd9d3d6 100644
--- a/Makefile.shared
+++ b/Makefile.shared
@@ -7,11 +7,11 @@
 
 # CC contains the current compiler.  This one MUST be defined
 CC=cc
-CFLAGS=$(CFLAG)
+CFLAGS=$(CFLAG) -march=armv7-a -mfloat-abi=softfp
 # LDFLAGS contains flags to be used when temporary object files (when building
 # shared libraries) are created, or when an application is linked.
 # SHARED_LDFLAGS contains flags to be used when the shared library is created.
-LDFLAGS=
+LDFLAGS=-Wl,--fix-cortex-a8
 SHARED_LDFLAGS=
 
 NM=nm
diff --git a/conf b/conf
new file mode 100755
index 0000000..133bb52
--- /dev/null
+++ b/conf
@@ -0,0 +1,59 @@
+./Configure \
+  no-md2 \
+  no-md4 \
+  no-mdc2 \
+  no-ripemd \
+  no-whrlpool \
+  no-des \
+  no-rc2 \
+  no-rc4 \
+  no-rc5 \
+  no-idea \
+  no-cast \
+  no-camellia \
+  no-seed \
+  no-cms \
+  no-ec \
+  no-rsa \
+  no-dsa \
+  no-ecdsa \
+  no-dh \
+  no-ecdh \
+  no-x509v3 \
+  no-dso \
+  no-engine \
+  no-pem \
+  no-x509 \
+  no-x509v3 \
+  no-conf \
+  no-txt_db \
+  no-pkcs7 \
+  no-pkcs12 \
+  no-comp \
+  no-ocsp \
+  no-ui \
+  no-krb5 \
+  no-srp \
+  no-store \
+  no-cmac \
+  no-bf \
+  no-cms \
+  no-pqueue \
+  no-ts \
+  no-jpake \
+  no-md5 \
+  no-dso \
+  -DOPENSSL_NO_PKEY \
+  -DOPENSSL_NO_SOCK \
+  -DOPENSSL_NO_SCTP \
+  -DOPENSSL_NO_DGRAM \
+  -DOPENSSL_NO_EC2M \
+  -DOPENSSL_NO_SHA0 \
+  -DOPENSSL_NO_SHA512 \
+  -DOPENSSL_NO_SHA256 \
+  -DOPENSSL_NO_RFC3779 \
+  -DOPENSSL_NO_DEPRECATED \
+  -DOPENSSL_NO_MULTIBLOCK \
+  -DOPENSSL_NO_OBJECT \
+  android-armv7
+
diff --git a/crypto/evp/digest.c b/crypto/evp/digest.c
index bd7760d..9429486 100644
--- a/crypto/evp/digest.c
+++ b/crypto/evp/digest.c
@@ -224,6 +224,7 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
 #ifndef OPENSSL_NO_ENGINE
  skip_to_init:
 #endif
+#ifndef OPENSSL_NO_PKEY
     if (ctx->pctx) {
         int r;
         r = EVP_PKEY_CTX_ctrl(ctx->pctx, -1, EVP_PKEY_OP_TYPE_SIG,
@@ -231,6 +232,7 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
         if (r <= 0 && (r != -2))
             return 0;
     }
+#endif
     if (ctx->flags & EVP_MD_CTX_FLAG_NO_INIT)
         return 1;
 #ifdef OPENSSL_FIPS
@@ -328,6 +330,7 @@ int EVP_MD_CTX_copy_ex(EVP_MD_CTX *out, const EVP_MD_CTX *in)
 
     out->update = in->update;
 
+#ifndef OPENSSL_NO_PKEY
     if (in->pctx) {
         out->pctx = EVP_PKEY_CTX_dup(in->pctx);
         if (!out->pctx) {
@@ -335,6 +338,7 @@ int EVP_MD_CTX_copy_ex(EVP_MD_CTX *out, const EVP_MD_CTX *in)
             return 0;
         }
     }
+#endif
 
     if (out->digest->copy)
         return out->digest->copy(out, in);
@@ -384,8 +388,10 @@ int EVP_MD_CTX_cleanup(EVP_MD_CTX *ctx)
         OPENSSL_free(ctx->md_data);
     }
 #endif
+#ifndef OPENSSL_NO_PKEY
     if (ctx->pctx)
         EVP_PKEY_CTX_free(ctx->pctx);
+#endif
 #ifndef OPENSSL_NO_ENGINE
     if (ctx->engine)
         /*
-- 
1.9.5

